File: Code/Herkenningssystemen_hoofdstuk_4/ResNet50/PyTorch_4-2/PyTorch_Herkenning/app/src/main/java/com/example/pytorch_herkenning/MainActivity.java
package com.example.pytorch_herkenning;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import org.pytorch.IValue;
import org.pytorch.Module;
import org.pytorch.Tensor;
import org.pytorch.torchvision.TensorImageUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Calendar;

public class MainActivity extends AppCompatActivity {

    private static int RESULT_LOAD_IMAGE = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button buttonLoadImage = (Button) findViewById(R.id.button);
        Button detectButton = (Button) findViewById(R.id.detect);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[]  {android.Manifest.permission.READ_EXTERNAL_STORAGE}, 1);
        }

        buttonLoadImage.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View arg0) {
                TextView textView = findViewById(R.id.result_text);
                textView.setText("");
                Intent i = new Intent(Intent.ACTION_GET_CONTENT);
                i.setType("image/*");
                startActivityForResult(i, RESULT_LOAD_IMAGE);
            }
        });

        detectButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View arg0) {

                //Getting the image from the image view
                ImageView imageView = (ImageView) findViewById(R.id.image);
                Bitmap bitmap = ((BitmapDrawable)imageView.getDrawable()).getBitmap();
                Module module = null;
                try {
                    module = Module.load(assetFilePath(MainActivity.this, "model.pt"));
                } catch (IOException e) {
                    System.out.println("FAILED");
                    e.printStackTrace();
                }
                System.out.println(bitmap.getHeight());
                bitmap = Bitmap.createScaledBitmap(bitmap, 224, 224, true);

                //Input Tensor
               // float[] mean = new float[]{0.0f, 0.0f, 0.0f};
               // float[] std = new float[]{1.0f, 1.0f, 1.0f};
               // final Tensor inputTensor = TensorImageUtils.bitmapToFloat32Tensor(bitmap, mean, std);
                Tensor inputTensor = TensorImageUtils.bitmapToFloat32Tensor(bitmap,
                        TensorImageUtils.TORCHVISION_NORM_MEAN_RGB, TensorImageUtils.TORCHVISION_NORM_STD_RGB);

                Calendar currentTime = Calendar.getInstance();
                Tensor outputTensor = module.forward(IValue.from(inputTensor)).toTensor();
                System.out.println("INFERENCE TIME: " + (Calendar.getInstance().get(Calendar.MILLISECOND) - currentTime.get(Calendar.MILLISECOND)));

                float[] scores = outputTensor.getDataAsFloatArray();

                float maxScore = -Float.MAX_VALUE;
                int maxScoreIdx = -1;
                for (int i = 0; i < scores.length; i++) {
                    if (scores[i] > maxScore) {
                        maxScore = scores[i];
                        maxScoreIdx = i;
                    }
                }
                String className = ImageNetClasses.IMAGENET_CLASSES[maxScoreIdx];

                TextView textView = findViewById(R.id.result_text);
                System.out.println("res");
                textView.setText(className + " (" +maxScore+")");
            }
        });

    }
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        //This functions return the selected image from gallery
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == RESULT_LOAD_IMAGE && resultCode == RESULT_OK && null != data) {
            Uri selectedImage = data.getData();
            String[] filePathColumn = { MediaStore.Images.Media.DATA };

            Cursor cursor = getContentResolver().query(selectedImage,
                    filePathColumn, null, null, null);
            cursor.moveToFirst();

            int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
            String picturePath = cursor.getString(columnIndex);
            cursor.close();

            ImageView imageView = (ImageView) findViewById(R.id.image);
            imageView.setImageBitmap(BitmapFactory.decodeFile(picturePath));
            System.out.println(imageView);
            imageView.setImageURI(selectedImage);
        }
    }

    public static String assetFilePath(Context context, String assetName) throws IOException {
        File file = new File(context.getFilesDir(), assetName);
        if (file.exists() && file.length() > 0) {
            return file.getAbsolutePath();
        }

        try (InputStream is = context.getAssets().open(assetName)) {
            try (OutputStream os = new FileOutputStream(file)) {
                byte[] buffer = new byte[4 * 1024];
                int read;
                while ((read = is.read(buffer)) != -1) {
                    os.write(buffer, 0, read);
                }
                os.flush();
            }
            return file.getAbsolutePath();
        }
    }
}

File: Code/Herkenningssystemen_hoofdstuk_4/ResNet50/Onnx_android/ONNX_Herkenning/app/src/main/java/com/example/onnx_herkenning/MainActivity.kt
package com.example.onnx_herkenning

//import com.example.onnx_runtime.R
//import com.example.onnx_runtime.MainActivity
import ai.onnxruntime.*
import android.Manifest
import android.content.Intent
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.os.Build
import android.os.Bundle
import android.provider.MediaStore
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.nio.FloatBuffer
import java.util.*

const val DIM_BATCH_SIZE = 1;
const val DIM_PIXEL_SIZE = 3;
const val IMAGE_SIZE_X = 224;
const val IMAGE_SIZE_Y = 224;
const val IMAGE_MEAN: Float = 0.0f;
const val IMAGE_STD: Float = 1.0f;

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val buttonLoadImage = findViewById<View>(R.id.button) as Button
        val detectButton = findViewById<View>(R.id.detect) as Button

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), 1)
        }

        buttonLoadImage.setOnClickListener {
            val textView = findViewById<TextView>(R.id.result_text)
            textView.text = ""
            val i = Intent(
                Intent.ACTION_PICK,
                MediaStore.Images.Media.EXTERNAL_CONTENT_URI
            )
            startActivityForResult(i, RESULT_LOAD_IMAGE)
        }

        detectButton.setOnClickListener {
            println("INITIALIZE")
            var env = OrtEnvironment.getEnvironment()

            // Laad het correcte ONNX model
            var session = env.createSession(resources.openRawResource(R.raw.model_tf).readBytes())
            val labellist = loadLabelList()

            println("PREPARE DATA")
            val imageView = findViewById<View>(R.id.image) as ImageView
            var bitmap = (imageView.drawable as BitmapDrawable).bitmap

            // ONNX PyTorch input formaat
             //val shape = longArrayOf(1, 3, IMAGE_SIZE_X.toLong(), IMAGE_SIZE_Y.toLong())

            // ONNX Tensorflow input formaat
            val shape = longArrayOf(1, IMAGE_SIZE_X.toLong(), IMAGE_SIZE_Y.toLong(),3)

            bitmap = Bitmap.createScaledBitmap(bitmap, IMAGE_SIZE_X, IMAGE_SIZE_Y, true)

            // Preprocess voor Pytorch ONNX model
             //val imgData = preprocessPy(bitmap)

            // Preprocess voor TensorFlow ONNX model
            val imgData = preprocessTf(bitmap)

            var t1 = OnnxTensor.createTensor(env, imgData, shape)

            println("EXECUTE")
            //var inputs = Collections.singletonMap("input", t1)
            var inputs = Collections.singletonMap("input_1", t1)
            val currentTime = Calendar.getInstance()
            val output = session?.run(inputs)
            println("INFERENCE TIME: " + (Calendar.getInstance()[Calendar.MILLISECOND] - currentTime[Calendar.MILLISECOND]))


            println("PROCESS OUTPUT")
            val labelVals = output?.get(0)?.value as Array<FloatArray>
            val probabilities = labelVals[0]

            var maxScore = -100000f
            var maxScoreIdx = -1
            for (i in 0 until probabilities.size) {
                if (probabilities[i] > maxScore) {
                    maxScore = probabilities[i]
                    maxScoreIdx = i
                }
            }
            val className = labellist?.get(maxScoreIdx)
            val textView = findViewById<TextView>(R.id.result_text)
            textView.text = className + "(" + maxScore + ")"

            println("FINISHED")
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        //This functions return the selected image from gallery
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RESULT_LOAD_IMAGE && resultCode == RESULT_OK && null != data) {
            val selectedImage = data.data
            val filePathColumn = arrayOf(MediaStore.Images.Media.DATA)
            val cursor = contentResolver.query(
                selectedImage!!,
                filePathColumn, null, null, null
            )
            cursor!!.moveToFirst()
            val columnIndex = cursor.getColumnIndex(filePathColumn[0])
            val picturePath = cursor.getString(columnIndex)
            cursor.close()
            val imageView = findViewById<View>(R.id.image) as ImageView
            imageView.setImageBitmap(BitmapFactory.decodeFile(picturePath))

            imageView.setImageURI(selectedImage)
        }
    }

    companion object {
        private const val RESULT_LOAD_IMAGE = 1
    }

    @Throws(IOException::class)
    private fun loadLabelList(): List<String>? {
        var labelList: List<String> = ArrayList()
        val reader = BufferedReader(InputStreamReader(this.assets.open("labels.txt")))
        labelList = reader.readLines()
        reader.close()
        return labelList
    }
}

fun preprocessPy(bitmap: Bitmap): FloatBuffer {
    val imgData = FloatBuffer.allocate(
        DIM_BATCH_SIZE
                * IMAGE_SIZE_X
                * IMAGE_SIZE_Y
                * DIM_PIXEL_SIZE
    )
    imgData.rewind()

    val stride = IMAGE_SIZE_X * IMAGE_SIZE_Y
    val bmpData = IntArray(IMAGE_SIZE_X * IMAGE_SIZE_Y)
    bitmap.getPixels(bmpData, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)

    var idx: Int = 0
    for (i in 0..IMAGE_SIZE_X - 1) {
        for (j in 0..IMAGE_SIZE_Y - 1) {
            val idx = IMAGE_SIZE_Y * i + j
            val pixelValue = bmpData[idx]
            imgData.put(idx, (((pixelValue shr 16 and 0xFF) / 255f - 0.485f) / 0.229f))
            imgData.put(idx + stride, (((pixelValue shr 8 and 0xFF) / 255f - 0.456f) / 0.224f))
            imgData.put(idx + stride * 2, (((pixelValue and 0xFF) / 255f - 0.406f) / 0.225f))
        }
    }

    imgData.rewind()
    return imgData
}

fun preprocessTf(bitmap: Bitmap): FloatBuffer {
    val imgData = FloatBuffer.allocate(
        DIM_BATCH_SIZE
                * IMAGE_SIZE_X
                * IMAGE_SIZE_Y
                * DIM_PIXEL_SIZE)
    imgData.rewind()

    val bmpData = IntArray(IMAGE_SIZE_X * IMAGE_SIZE_Y)
    bitmap.getPixels(bmpData, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)

    var idx: Int = 0
    for (i in 0..IMAGE_SIZE_X - 1) {
        for (j in 0..IMAGE_SIZE_Y - 1) {
            val pixelValue = bmpData[idx++]
            imgData.put(((pixelValue shr 16 and 0xFF) - IMAGE_MEAN) / IMAGE_STD)
            imgData.put(((pixelValue shr 8 and 0xFF) - IMAGE_MEAN) / IMAGE_STD)
            imgData.put(((pixelValue and 0xFF) - IMAGE_MEAN) / IMAGE_STD)
        }
    }

    imgData.rewind()
    return imgData
}


